[tools][web] Make Plugin Registrant file ephemeral.

ditman opened the pull request:
This change modifies flutter_tool slightly to make the `web_plugin_registrant.dart` file ephemeral (only created when needed, and not left in the users' working copy).

I needed to do some changes to the way a FlutterWeb Project works so it supports both modes of operation (on-disk for `flutter build` and `in-memory` for `flutter run`) so we can configure where the `flutter_plugins.dart` code creates the plugin registrar file for the web. Hence why it'd be nice that @blasten and @stuartmorgan took a quick look if possible :)

### PluginMigrator

As suggested by @stuartmorgan, this change also adds a `ProjectMigrator` (class `ScrubGeneratedPluginRegistrant`) that deletes the old `lib/generated_plugin_registrant.dart` file if it exists, and cleans the `.gitignore` file by removing lines that are equal to: `'lib/generated_plugin_registrant.dart'`.

The migrator runs when the user does `flutter build web` (only), with the following output:

```
$ flutter build web
Running "flutter pub get" in flutter_example...                  1,041ms

💪 Building with sound null safety 💪

generated_plugin_registrant.dart found. Deleted.
Upgrading .gitignore
Compiling lib/main.dart for the Web...                             8.5s
```

### More changes (cleanup)

* Creates a `file_generators` subdir, and moves there the file that generates `flutter.js`
  * Unifies the code that generates the main entrypoint of a web app (and puts it in a file under `file_generators`)
  * Extracts the code that generates the service worker (and puts it in a file under `file_generators`)
* Prevents the service worker from prefetching `assets/NOTICES`. [Issue](https://github.com/flutter/flutter/issues/101828).
* Simplifies the web entrypoints a little bit (`main.dart` and `web_plugin_registrar.dart`), and ensures they won't trigger `lint` warnings. [Issue](https://github.com/flutter/flutter#99952).
* Removes `generated_plugin_registrar.dart` from `.gitignore` template for new projects, now that it is created on-demand.

### Issues

* Fixes https://github.com/flutter/flutter/issues/101828
* Fixes https://github.com/flutter/flutter/issues/99952
* Somewhat opinates on https://github.com/flutter/flutter/issues/80406 (doesn't merge the files)

### Tests

WIP, this PR has been created in part to see where things burst at the seams :)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ditman committed:
Move flutter_js.dart to file_generators subdir.

Fix imports.

ditman committed:
Unify generation of web's main.dart.

Fix flutter build + flutter run scripts to use the new generation
function, instead of their hardcoded versions.

ditman committed:
Split flutter_service_worker_js to its own file.

Fixes https://github.com/flutter/flutter/issues/101828

ditman committed:
Make web_plugin_registrant.dart ephemeral

ditman committed:
Add ignore_for_file:type=lint to web entrypoints.

(This shouldn't be needed, since the generated files live outside of the
main app's lib directory, and are only created when doing flutter
run/build, but it's good practice that generated files don't interfere
with user code anyway.)

Fixes https://github.com/flutter/flutter/issues/99952

ditman committed:
Remove generated_plugin_registrant from gitignore.

The file is now ephemeral, and generated only as needed, so there's no
need for this gitignore to exist anymore.

ditman committed:
Do not expect generated_plugin_registrant to exist anymore.

ditman committed:
Do not call injectPlugins on web without a configured project.

Prevent ensureReadyForPlatformSpecificTooling from calling injectPlugins
for the web, since at this point on the build we don't have enough info.
Flutter web commands (build, run) know when to call injectPlugins.

ditman committed:
Do not call injectPlugins for web without configuring it first.

ditman committed:
Fix expectations about generated files.

ditman committed:
Do not expect that assets/NOTICES is prefetched

ditman committed:
Restore _NullaryFunction (required for some tests to pass)

ditman committed:
Fix web plugin registrant integration tests.

stuartmorgan commented on packages/flutter_tools/lib/src/flutter_plugins.dart:
```diff
@@ -1145,6 +1148,7 @@ Future<void> injectPlugins(
     }
   }
   if (webPlatform) {
+    assert(project.web.buildDir != null, 'project.web.buildDir must be set before calling injectPlugins');
```
If web plugin injection is now fundamentally different than on other platforms (because it's per-build), we should probably split it out into a separate method; it seems like we're probably never going to actually want to call this for web at the same time as other platforms, making this a foot-gun. (Thus the need for the commented out line with explanation in the other file.) So, `injectPersistentPluginFiles` and `injectBuildTimePluginFiles`?

Or alternately, a required `injectionTime` enum argument to this single method with `setup` and `build`, and web can internally no-op when injectionTime is not `build`.

ditman replied:
Yes, this is a good call. The problem is that injectPlugins is called from another method that is called "ensureReadyForPlatformSpecificTooling" which seems interesting even for web (it's delegating to a sub-method of a class). The problem is that it then calls to injectPlugins, potentially for many platforms at the same time. I might remove *web* from injectPlugins at all, and just have a custom method.

ditman replied:
I've removed the `webPlatform` option from `injectPlugins`, and created a separate `injectBuildTimePluginFiles` method (almost identical to `injectPlugins` in signature) that lets the user inject plugin files as needed (where needed), for now only for webPlatform.

ditman commented on packages/flutter_tools/lib/src/web/compile.dart:
```diff
@@ -32,7 +32,6 @@ Future<void> buildWeb(
   final Directory outputDirectory = globals.fs.directory(getWebBuildDirectory());
   outputDirectory.createSync(recursive: true);
 
-  await injectPlugins(flutterProject, webPlatform: true);
```
Yes, this was a duplicate with line 38: `globals.buildSystem.build(WebServiceWorker`... which eventually called `injectPlugins` again through the following "trace":

```
buildWeb
buildSystem.build (WebServiceWorker Target)
... WebServiceWorkerTarget depends on: Dart2JSTarget, WebReleaseBundle, WebBuiltInAssets)
... Dart2JSTarget depends on WebEntrypointTarget, GenerateLocalizationsTarget
... WebEntrypointTarget
build -> await injectPlugins (line 124 of lib/src/build_system/targets/web.dart)
```

ditman replied:
However at the time that this is called, it's still not "clear" where in the filesystem is the build going to end up. The "directory" in which the build is going to end up (in-memory or physical) is created (as a random temp thing) somewhere along the trace (and also on flutter run vs build).

ditman replied:
This location was useful however, to hook up the ProjectMigrators for web!

ditman commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -748,6 +751,12 @@ class WebProject extends FlutterProjectPlatform {
 
+  /// The directory where the build is writing files.
+  ///
+  /// This needs to be set by the build scripts, because it depends of what
```
Good one, thanks!

ditman commented on packages/flutter_tools/lib/src/flutter_plugins.dart:
```diff
@@ -961,9 +959,14 @@ Future<void> _writeWebPluginRegistrant(FlutterProject project, List<Plugin> plug
+
   if (webPlugins.isEmpty) {
     ErrorHandlingFileSystem.deleteIfExists(pluginFile);
+    // Render noop plugin registrant file
```
Ah yes, I forgot to implement this, just creating an empty noop plugin registrant file in case there are no plugins, so we don't have to special-case down the line.

ditman replied:
I've added a noop registerPlugins() {} function so the web_plugins_registrar.dart file can be generated (and used down the line) unconditionally.

ditman committed:
Separate web platform from injectPlugins method.

* Remove webPlatform from injectPlugins, to remove a foot gun from the
  execution path.
* Introduce injectBuildTimePluginFiles with a similar signature to
  injectPlugins so web (and maybe others?) can inject the plugin files
  it needs, when needed.
* Always inject a web plugin registrant, either a NOOP, or something
  that actually registers plugins, to decrease tooling complexity down
  the line (when generating a main entrypoint, for example)
* Cleanup of other unneeded things (like the buildDir property on the
  WebProject, that is now an argument passed to
  injectBuildTimePluginFiles.
* Test injectBuildTimePluginFiles.

ditman commented:
I seem to recall that I have at least two tests that test for the *non-existence* of the web_plugin_registrant, but can't find them right now... They'll eventually explode in CI, I guess :)

yjbanov commented:
I am not too familiar with the nuances of plugin registration logic, so while this PR LGTM I recommend getting LGTMs from @stuartmorgan and/or @christopherfujino.

ditman committed:
Override the TemplateRenderer in tests now that the web template is written unconditionally.

ditman commented on packages/flutter_tools/lib/src/flutter_plugins.dart:
```diff
@@ -961,9 +959,14 @@ Future<void> _writeWebPluginRegistrant(FlutterProject project, List<Plugin> plug
   final Map<String, Object> context = <String, Object>{
     'methodChannelPlugins': webPlugins,
   };
-  final File pluginFile = project.web.libDirectory.childFile('generated_plugin_registrant.dart');
```
I've never heard of this before, is this a migration guide (like [this](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects)) or something that can be automated?

ditman replied:
Looking at this https://github.com/flutter/flutter/pull/100647

ditman replied:
I've added a ProjectMigrator for web, and hooked it up where the script (used) to create the file when doing a `flutter build`. Added a test for it too.

ditman committed:
Add ProjectMigrator to clean old refs to generated_plugin_registrant, and hook it up at the beginning of the compile process, where the file used to be generated.

ditman commented:
Addressed PR comments:

* Added a ProjectMigrator to scrub `generated_plugin_registrant.dart`
* Separated `injectPlugins` from `injectBuildTimePluginFiles`
  * `injectPlugins` does not support `webPlatform` anymore.
  * `injectBuildTimePluginFiles` has a signature very similar to `injectPlugins`, in case it can be reused for other platforms. It **only** supports `webPlatform` for now.
* Create a noop `web_plugin_registrar.dart` when the app has no plugins, so the rest of the build process doesn't need to care about the project having plugins or not.

I hope I've addressed your comments @stuartmorgan, PTAL!

ditman committed:
Small fix to docs post-rebase

ditman commented:
(Rebased after https://github.com/flutter/flutter/commit/944fcda67f294981dfe2b57991c32afc724f34c2)

ditman commented:
PTAL @stuartmorgan? Thanks!

stuartmorgan commented on packages/flutter_tools/lib/src/flutter_plugins.dart:
```diff
@@ -1068,8 +1077,35 @@ Future<void> refreshPluginsList(
+/// only at build-time.
+///
+/// This method is similar to [injectPlugins], but used only for platforms where
+/// the plugin files are not required at compile-time (currently: Web).
```
Don't you mean not required at project generation time? If not, I'm not sure what the distinction between "compile time" and "build time" is.

ditman replied:
Yeah, I think I'm going to use "project creation time" and "build time", I can't think of better names :/

ditman replied:
I also specified a little bit better what the *actual* difference between the two methods is:

* one writes the generated plugin registrants in locations configured by each `FlutterProject` subclass,
* the other writes the generated plugin registrant (later) and within the temp directory of the build in progress (maybe to a real filesystem, or to an in-memory one)

stuartmorgan commented on packages/flutter_tools/lib/src/flutter_plugins.dart:
```diff
@@ -1068,8 +1077,35 @@ Future<void> refreshPluginsList(
 ///
+/// These files are required at compile-time.
+///
+/// This method used to generate the plugin registrar for the webPlatform too,
```
Rather than say what used to happen (which is probably not interesting to most readers), I would just focus on the explanation of why web isn't here.

ditman replied:
I rewrote these two bits, I think the difference is more clear now (or somebody who's actually reading these docs will have a better context of what's going on)

stuartmorgan-g reviewed (approved):
LGTM with comment nits, with the caveat that I'm not very familiar with how the web bootstrapping works.

Jasguerrero reviewed (approved):
LGTM

ditman committed:
Improve docs in the injectPlugins/injectBuildTimePluginFiles directories

ditman commented:
(Applying the correct tag this time 😅)
